util/dvsim/Deploy.py

```diff
@@ -279,118 +277,138 @@ def dispatch_cmd(self):
 
     def _test_passed(self):
-        '''Return True if the job passed, False otherwise
+        """Determine the outcome of the job (P/F if it ran to completion).
 
-        This is called by poll() just after the job finishes.
+        Return True if the job passed, False otherwise. This is called by
+        poll() just after the job finishes.
+
+        """
 
-        '''
         if self.dry_run:
             return True
 
-        seen_fail_pattern = False
-        for fail_pattern in self.fail_patterns:
-            # Return error message with the following 4 lines.
-            grep_cmd = "grep -m 1 -A 4 -E \'" + fail_pattern + "\' " + self.log
-            (status, rslt) = subprocess.getstatusoutput(grep_cmd)
-            if rslt:
-                msg = "```\n{}\n```\n".format(rslt)
-                self.fail_msg += msg
-                log.log(VERBOSE, msg)
-                seen_fail_pattern = True
-                break
-
-        if seen_fail_pattern:
+        # Only one pattern need to be seen.
+        failed = False
+        chk_failed = bool(self.fail_patterns)
+
+        # All pass patterns need to be seen - hence maintain it as bool dict.
+        passed = {p: False for p in self.pass_patterns}
+        chk_passed = bool(self.pass_patterns)
+
+        try:
+            f = open(self.log, "r", encoding="UTF-8")
+        except Exception as e:
+            msg = "Error opening file \"{!r}\":\n{!r}".format(self.log, e)
+            self.fail_msg += msg
+            log.log(VERBOSE, msg)
+            return False
+
+        lines = f.readlines()
+
+        def _find_patterns(patterns, line):
+            '''Helper function that returns true if all or any of the given
+            patterns is found, else False.'''
+
+            assert patterns
+            for pattern in patterns:
+                match = re.search(r"{}".format(pattern), line)
+                if match:
+                    return pattern
+            return None
+
+        def _get_n_lines(pos, num):
+            "Helper function that returns next N lines starting at pos index."
+
+            idx = min(len(lines) - 1, pos + num - 1)
+            return ''.join(lines[pos:idx]).strip()
+
+        if chk_failed or chk_passed:
+            for cnt, line in enumerate(lines):
+                if chk_failed:
+                    if _find_patterns(self.fail_patterns, line) is not None:
+                        # Print 4 additional lines to help debug more easily.
+                        msg = "```\n{}\n```\n".format(_get_n_lines(cnt, 5))
+                        self.fail_msg += msg
+                        log.log(VERBOSE, msg)
+                        failed = True
+                        chk_failed = False
+
+                if chk_passed:
+                    pattern = _find_patterns(self.pass_patterns, line)
+                    if pattern is not None:
+                        passed[pattern] = True
+                        chk_passed = not all(passed.values())
```

Comment:
This is neat, but it might be easier to remove the pass pattern that we match? That way, later calls to `_find_patterns` don't have to search for it.